## Dataset Summary

[**Global Coffee Health Dataset**](https://www.kaggle.com/datasets/uom190346a/global-coffee-health-dataset):

The GlobalCoffeeHealth dataset contains 10,000 synthetic records reflecting real-world patterns of coffee consumption, sleep behavior, and health outcomes across 20 countries. It includes demographics, daily coffee intake, caffeine levels, sleep duration and quality, BMI, heart rate, stress, physical activity, health issues, occupation, smoking, and alcohol consumption.

The dataset captures realistic correlations observed in research—such as caffeine’s impact on sleep, stress, and health—making it ideal for statistical analysis, predictive modeling, and lifestyle or wellness studies.


*   **Scope**: Individual-level records linking coffee/caffeine habits, sleep, vitals, lifestyle, and self-reported health across multiple countries.
*   **Shape**: 10,000 rows × 16 columns
*   **Granularity**: One row per person (no time series).
*   **Keys**: ID is unique per row.
*   **Missingness**: None observed.
*   **Types**: Numeric (continuous + integer), categorical, and binary (0/1).




---



**Data Description**:

| Column                   | Type       | Description                                                   |
|---------------------------|------------|---------------------------------------------------------------|
| ID                        | Integer    | Unique record ID (1–10000)                                   |
| Age                       | Integer    | Age of participant (18–80 years)                             |
| Gender                    | Categorical| Male, Female, Other                                          |
| Country                   | Categorical| Country of residence (20 countries)                          |
| Coffee_Intake             | Float      | Daily coffee consumption in cups (0–10)                      |
| Caffeine_mg               | Float      | Estimated daily caffeine intake in mg (1 cup ≈ 95 mg)        |
| Sleep_Hours               | Float      | Average hours of sleep per night (3–10 hours)                |
| Sleep_Quality             | Categorical| Poor, Fair, Good, Excellent (based on sleep hours)           |
| BMI                       | Float      | Body Mass Index (15–40)                                      |
| Heart_Rate                | Integer    | Resting heart rate (50–110 bpm)                              |
| Stress_Level              | Categorical| Low, Medium, High (based on sleep hours and lifestyle)       |
| Physical_Activity_Hours   | Float      | Weekly physical activity (0–15 hours)                        |
| Health_Issues             | Categorical| None, Mild, Moderate, Severe (based on age, BMI, and sleep)  |
| Occupation                | Categorical| Office, Healthcare, Student, Service, Other                  |
| Smoking                   | Boolean    | 0 = No, 1 = Yes                                              |
| Alcohol_Consumption       | Boolean    | 0 = No, 1 = Yes                                              |


# Importing Libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import LabelEncoder, OrdinalEncoder, StandardScaler

# Load Dataset

In [ ]:
df = pd.read_csv('/content/synthetic_coffee_health_10000.csv')

# Exploring the Dataset

In [ ]:
df

,ID,Age,Gender,Country,Coffee_Intake,Caffeine_mg,Sleep_Hours,Sleep_Quality,BMI,Heart_Rate,Stress_Level,Physical_Activity_Hours,Health_Issues,Occupation,Smoking,Alcohol_Consumption
0,1,40,Male,Germany,3.5,328.1,7.5,Good,24.9,78,Low,14.5,NaN,Other,0,0
1,2,33,Male,Germany,1.0,94.1,6.2,Good,20.0,67,Low,11.0,NaN,Service,0,0
2,3,42,Male,Brazil,5.3,503.7,5.9,Fair,22.7,59,Medium,11.2,Mild,Office,0,0
3,4,53,Male,Germany,2.6,249.2,7.3,Good,24.7,71,Low,6.6,Mild,Other,0,0
4,5,32,Female,Spain,3.1,298.0,5.3,Fair,24.1,76,Medium,8.5,Mild,Student,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,9996,50,Female,Japan,2.1,199.8,6.0,Fair,30.5,50,Medium,10.1,Moderate,Healthcare,0,1
9996,9997,18,Female,UK,3.4,319.2,5.8,Fair,19.1,71,Medium,11.6,Mild,Service,0,0
9997,9998,26,Male,China,1.6,153.4,7.1,Good,25.1,66,Low,13.7,NaN,Student,1,1
9998,9999,40,Female,Finland,3.4,327.1,7.0,Good,19.3,80,Low,0.1,NaN,Student,0,0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       10000 non-null  int64  
 1   Age                      10000 non-null  int64  
 2   Gender                   10000 non-null  object 
 3   Country                  10000 non-null  object 
 4   Coffee_Intake            10000 non-null  float64
 5   Caffeine_mg              10000 non-null  float64
 6   Sleep_Hours              10000 non-null  float64
 7   Sleep_Quality            10000 non-null  object 
 8   BMI                      10000 non-null  float64
 9   Heart_Rate               10000 non-null  int64  
 10  Stress_Level             10000 non-null  object 
 11  Physical_Activity_Hours  10000 non-null  float64
 12  Health_Issues            4059 non-null   object 
 13  Occupation               10000 non-null  object 
 14  Smoking                

**Data types include**:

*   int64: 5 columns
*   float64: 5 columns
*   object (categorical/text): 6 columns

**Missing values**: In Health_Issues there is 5941 row missing.


In [ ]:
df.describe()

,ID,Age,Coffee_Intake,Caffeine_mg,Sleep_Hours,BMI,Heart_Rate,Physical_Activity_Hours,Smoking,Alcohol_Consumption
count,10000.00000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.00000,10000.000000
mean,5000.50000,34.949100,2.509230,238.411010,6.636220,23.986860,70.617800,7.48704,0.20040,0.300700
std,2886.89568,11.160939,1.450248,137.748815,1.222055,3.906411,9.822951,4.31518,0.40032,0.458585
min,1.00000,18.000000,0.000000,0.000000,3.000000,15.000000,50.000000,0.00000,0.00000,0.000000
25%,2500.75000,26.000000,1.500000,138.750000,5.800000,21.300000,64.000000,3.70000,0.00000,0.000000
50%,5000.50000,34.000000,2.500000,235.400000,6.600000,24.000000,71.000000,7.50000,0.00000,0.000000
75%,7500.25000,43.000000,3.500000,332.025000,7.500000,26.600000,77.000000,11.20000,0.00000,1.000000
max,10000.00000,80.000000,8.200000,780.300000,10.000000,38.200000,109.000000,15.00000,1.00000,1.000000


**For Coffee_Intake column**:

*   **Count**: Number of non-missing is 10000 so there is no NULL values.
*   **Mean**: Average of all values is 2.5.
*   **std**: standard deviation that measuring how spread out the values are is 1.4.
*   **Min**: Minimum value in the column is 0.
*   **25%(Q1)**: The value below which 25% of the data falls is 1.5.
*   **50%(Q2, Median)**: The middle value of the column is 2.5.
*   **25%(Q3)**: The value below which 75% of the data falls is 3.5.
*   **Max**: The maximum value in the column is 8.2.

In [ ]:
df.describe(include='object')

,Gender,Country,Sleep_Quality,Stress_Level,Health_Issues,Occupation
count,10000,10000,10000,10000,4059,10000
unique,3,20,4,3,3,5
top,Female,Canada,Good,Low,Mild,Office
freq,5001,543,5637,6989,3579,2073


**For Sleep_Quality column**:

*   **Count**: Number of non-missing is 10000 so there is no NULL values.
*   **Unique**: Number of distinct values is 4 (Good, Fair, Excellent, Poor).
*   **Top**: most frequent value is Good.
*   **Freq**: frequency of that most frequent value is 5637.

In [ ]:
df = df.drop('ID', axis=1)

In [ ]:
df

,Age,Gender,Country,Coffee_Intake,Caffeine_mg,Sleep_Hours,Sleep_Quality,BMI,Heart_Rate,Stress_Level,Physical_Activity_Hours,Health_Issues,Occupation,Smoking,Alcohol_Consumption
0,40,Male,Germany,3.5,328.1,7.5,Good,24.9,78,Low,14.5,NaN,Other,0,0
1,33,Male,Germany,1.0,94.1,6.2,Good,20.0,67,Low,11.0,NaN,Service,0,0
2,42,Male,Brazil,5.3,503.7,5.9,Fair,22.7,59,Medium,11.2,Mild,Office,0,0
3,53,Male,Germany,2.6,249.2,7.3,Good,24.7,71,Low,6.6,Mild,Other,0,0
4,32,Female,Spain,3.1,298.0,5.3,Fair,24.1,76,Medium,8.5,Mild,Student,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,50,Female,Japan,2.1,199.8,6.0,Fair,30.5,50,Medium,10.1,Moderate,Healthcare,0,1
9996,18,Female,UK,3.4,319.2,5.8,Fair,19.1,71,Medium,11.6,Mild,Service,0,0
9997,26,Male,China,1.6,153.4,7.1,Good,25.1,66,Low,13.7,NaN,Student,1,1
9998,40,Female,Finland,3.4,327.1,7.0,Good,19.3,80,Low,0.1,NaN,Student,0,0


**Drop ID Column**: because it doesn’t carry any useful information about the data itself, it’s just a row identifier, not a feature that describes trends, patterns, or relationships.

# Missing Values

In [ ]:
df.isnull().sum()

,0
Age,0
Gender,0
Country,0
Coffee_Intake,0
Caffeine_mg,0
Sleep_Hours,0
Sleep_Quality,0
BMI,0
Heart_Rate,0
Stress_Level,0


In [ ]:
df['Health_Issues'].unique()

array([nan, 'Mild', 'Moderate', 'Severe'], dtype=object)

In [ ]:
df['Health_Issues'].value_counts()

,count
Health_Issues,
Mild,3579
Moderate,463
Severe,17


In [ ]:
df[df['Health_Issues'].isnull()]

,Age,Gender,Country,Coffee_Intake,Caffeine_mg,Sleep_Hours,Sleep_Quality,BMI,Heart_Rate,Stress_Level,Physical_Activity_Hours,Health_Issues,Occupation,Smoking,Alcohol_Consumption
0,40,Male,Germany,3.5,328.1,7.5,Good,24.9,78,Low,14.5,NaN,Other,0,0
1,33,Male,Germany,1.0,94.1,6.2,Good,20.0,67,Low,11.0,NaN,Service,0,0
5,32,Male,Mexico,3.4,326.4,6.4,Good,27.0,82,Low,8.8,NaN,Service,0,0
8,29,Male,UK,1.7,162.0,7.1,Good,21.7,60,Low,2.2,NaN,Service,1,1
10,29,Female,Switzerland,4.5,427.5,8.1,Excellent,21.5,66,Low,8.4,NaN,Other,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9992,42,Other,Spain,2.8,262.0,6.6,Good,16.8,77,Low,4.3,NaN,Service,0,1
9993,18,Female,Italy,0.0,0.0,6.2,Good,25.9,75,Low,8.3,NaN,Service,0,0
9997,26,Male,China,1.6,153.4,7.1,Good,25.1,66,Low,13.7,NaN,Student,1,1
9998,40,Female,Finland,3.4,327.1,7.0,Good,19.3,80,Low,0.1,NaN,Student,0,0


In [ ]:
df[df['Health_Issues'] =='Severe']

,Age,Gender,Country,Coffee_Intake,Caffeine_mg,Sleep_Hours,Sleep_Quality,BMI,Heart_Rate,Stress_Level,Physical_Activity_Hours,Health_Issues,Occupation,Smoking,Alcohol_Consumption
603,51,Male,UK,3.7,348.4,5.9,Fair,30.9,80,Medium,3.5,Severe,Service,0,0
931,57,Female,Italy,4.8,453.6,5.1,Fair,33.4,77,Medium,8.9,Severe,Other,0,0
1330,53,Male,Australia,0.0,0.0,5.5,Fair,30.5,70,Medium,2.8,Severe,Student,0,1
2838,61,Female,Japan,2.3,222.4,5.8,Fair,30.1,60,Medium,6.0,Severe,Service,0,0
3434,62,Female,Australia,0.5,48.1,5.0,Poor,30.8,82,High,1.9,Severe,Student,0,0
3977,52,Female,Italy,2.3,214.7,5.5,Fair,30.1,69,Medium,13.9,Severe,Student,1,0
4186,63,Female,USA,2.9,273.8,4.6,Poor,31.6,63,High,4.9,Severe,Student,1,0
4612,52,Male,Australia,5.2,490.2,5.8,Fair,31.1,73,Medium,3.8,Severe,Service,0,0
5036,61,Male,Finland,1.4,137.5,5.3,Fair,31.4,79,Medium,12.0,Severe,Other,0,1
5885,55,Female,Mexico,0.8,72.9,5.2,Fair,38.2,70,Medium,2.2,Severe,Service,0,1


In [ ]:
df['Health_Issues'] = df['Health_Issues'].fillna("No_Issue")

In [ ]:
df[df['Health_Issues'].isnull()]

,Age,Gender,Country,Coffee_Intake,Caffeine_mg,Sleep_Hours,Sleep_Quality,BMI,Heart_Rate,Stress_Level,Physical_Activity_Hours,Health_Issues,Occupation,Smoking,Alcohol_Consumption


**For Missing Values**:
- While analyzing the dataset, I found that only the Health_Issues column contains null values.  
- After checking the context, I discovered that these null values actually mean the person has no health issue.
- Therefore, I replaced the null values in Health_Issues with "No_Issue".


# Outliers

In [ ]:
numeric_features = df.select_dtypes(include=['int64', 'float64']).columns
categorical_features = df.select_dtypes(include='object').columns

In [ ]:
#Count of outliers

outlier_counts = {}
for col in numeric_features:
        Q1 = df[col].quantile(0.25)
        Q3 = df[col].quantile(0.75)
        IQR = Q3 - Q1
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR

        count = df[(df[col] < lower_bound) | (df[col] > upper_bound)].shape[0]
        outlier_counts[col] = count
outlier_counts

{'Age': 25,
 'Coffee_Intake': 39,
 'Caffeine_mg': 39,
 'Sleep_Hours': 26,
 'BMI': 40,
 'Heart_Rate': 50,
 'Physical_Activity_Hours': 0,
 'Smoking': 2004,
 'Alcohol_Consumption': 0}

In [ ]:
def remove_outliers_iqr(dataframe):
    df_clean = dataframe.copy()
    for col in numeric_features:
        Q1 = dataframe[col].quantile(0.25)
        Q3 = dataframe[col].quantile(0.75)
        IQR = Q3 - Q1
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR

        # Keep only rows within bounds for this column
        df_clean = df_clean[(df_clean[col] >= lower_bound) & (df_clean[col] <= upper_bound)]

    return df_clean

df = remove_outliers_iqr(df)
df

,Age,Gender,Country,Coffee_Intake,Caffeine_mg,Sleep_Hours,Sleep_Quality,BMI,Heart_Rate,Stress_Level,Physical_Activity_Hours,Health_Issues,Occupation,Smoking,Alcohol_Consumption
0,40,Male,Germany,3.5,328.1,7.5,Good,24.9,78,Low,14.5,No_Issue,Other,0,0
1,33,Male,Germany,1.0,94.1,6.2,Good,20.0,67,Low,11.0,No_Issue,Service,0,0
2,42,Male,Brazil,5.3,503.7,5.9,Fair,22.7,59,Medium,11.2,Mild,Office,0,0
3,53,Male,Germany,2.6,249.2,7.3,Good,24.7,71,Low,6.6,Mild,Other,0,0
4,32,Female,Spain,3.1,298.0,5.3,Fair,24.1,76,Medium,8.5,Mild,Student,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9993,18,Female,Italy,0.0,0.0,6.2,Good,25.9,75,Low,8.3,No_Issue,Service,0,0
9994,49,Female,Germany,1.6,150.1,5.7,Fair,25.5,81,Medium,12.9,Mild,Student,0,1
9995,50,Female,Japan,2.1,199.8,6.0,Fair,30.5,50,Medium,10.1,Moderate,Healthcare,0,1
9996,18,Female,UK,3.4,319.2,5.8,Fair,19.1,71,Medium,11.6,Mild,Service,0,0


In [ ]:
#Count of outliers after remove them

outlier_counts = {}
for col in numeric_features:
        Q1 = df[col].quantile(0.25)
        Q3 = df[col].quantile(0.75)
        IQR = Q3 - Q1
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR

        count = df[(df[col] < lower_bound) | (df[col] > upper_bound)].shape[0]
        outlier_counts[col] = count
outlier_counts

{'Age': 6,
 'Coffee_Intake': 0,
 'Caffeine_mg': 4,
 'Sleep_Hours': 0,
 'BMI': 0,
 'Heart_Rate': 0,
 'Physical_Activity_Hours': 0,
 'Smoking': 0,
 'Alcohol_Consumption': 0}

In [ ]:
df

,Age,Gender,Country,Coffee_Intake,Caffeine_mg,Sleep_Hours,Sleep_Quality,BMI,Heart_Rate,Stress_Level,Physical_Activity_Hours,Health_Issues,Occupation,Smoking,Alcohol_Consumption
0,40,Male,Germany,3.5,328.1,7.5,Good,24.9,78,Low,14.5,No_Issue,Other,0,0
1,33,Male,Germany,1.0,94.1,6.2,Good,20.0,67,Low,11.0,No_Issue,Service,0,0
2,42,Male,Brazil,5.3,503.7,5.9,Fair,22.7,59,Medium,11.2,Mild,Office,0,0
3,53,Male,Germany,2.6,249.2,7.3,Good,24.7,71,Low,6.6,Mild,Other,0,0
4,32,Female,Spain,3.1,298.0,5.3,Fair,24.1,76,Medium,8.5,Mild,Student,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9993,18,Female,Italy,0.0,0.0,6.2,Good,25.9,75,Low,8.3,No_Issue,Service,0,0
9994,49,Female,Germany,1.6,150.1,5.7,Fair,25.5,81,Medium,12.9,Mild,Student,0,1
9995,50,Female,Japan,2.1,199.8,6.0,Fair,30.5,50,Medium,10.1,Moderate,Healthcare,0,1
9996,18,Female,UK,3.4,319.2,5.8,Fair,19.1,71,Medium,11.6,Mild,Service,0,0


**For Outliers**:
- To check for outliers, I used the Interquartile Range (IQR) method.  
- The IQR is defined as the difference between the 75th percentile (Q3) and the 25th percentile (Q1).  
- Any value below Q1 - 1.5 * IQR or above Q3 + 1.5 * IQR is considered an outlier.  
- After detecting and count them, I remove them.


# Duplicated values

In [ ]:
df.duplicated().sum()

np.int64(0)

**There is no Duplicated values in the dataset**

# Encoding

In [ ]:
categorical_features

Index(['Gender', 'Country', 'Sleep_Quality', 'Stress_Level', 'Health_Issues',
       'Occupation'],
      dtype='object')

In [ ]:
for col in categorical_features:
    print(df[col].value_counts(),"\n")

Gender
Female    3912
Male      3775
Other      182
Name: count, dtype: int64 

Country
Canada         437
UK             419
Belgium        408
Finland        407
China          407
Norway         405
India          401
Germany        398
Spain          394
Australia      394
Switzerland    393
Netherlands    392
Italy          389
France         387
Sweden         386
Mexico         383
South Korea    373
Brazil         370
Japan          367
USA            359
Name: count, dtype: int64 

Sleep_Quality
Good         4428
Fair         1623
Excellent    1070
Poor          748
Name: count, dtype: int64 

Stress_Level
Low       5498
Medium    1623
High       748
Name: count, dtype: int64 

Health_Issues
No_Issue    4695
Mild        2821
Moderate     341
Severe        12
Name: count, dtype: int64 

Occupation
Office        1639
Other         1616
Healthcare    1550
Service       1539
Student       1525
Name: count, dtype: int64 



In [ ]:
df['Country'].unique()

array(['Germany', 'Brazil', 'Spain', 'Mexico', 'Switzerland',
       'Netherlands', 'Italy', 'UK', 'France', 'Belgium', 'Finland',
       'Australia', 'Canada', 'USA', 'Sweden', 'China', 'India', 'Japan',
       'Norway', 'South Korea'], dtype=object)

In [ ]:
# Label Encoding for Gender, Country and Occupation

label_encoder = LabelEncoder()

for col in ['Gender', 'Country', 'Occupation']:
    df[col] = label_encoder.fit_transform(df[col])

In [ ]:
# Ordinal Encoding for Sleep_Quality, Stress_Level, Health_Issues

from sklearn.preprocessing import OrdinalEncoder

sleep_encoder = OrdinalEncoder(categories=[['Poor', 'Fair', 'Good', 'Excellent']])
stress_encoder = OrdinalEncoder(categories=[['Low', 'Medium', 'High']])
health_encoder = OrdinalEncoder(categories=[['No_Issue', 'Mild', 'Moderate', 'Severe']])

df['Sleep_Quality'] = sleep_encoder.fit_transform(df[['Sleep_Quality']])
df['Stress_Level'] = stress_encoder.fit_transform(df[['Stress_Level']])
df['Health_Issues'] = health_encoder.fit_transform(df[['Health_Issues']])

In [ ]:
df

,Age,Gender,Country,Coffee_Intake,Caffeine_mg,Sleep_Hours,Sleep_Quality,BMI,Heart_Rate,Stress_Level,Physical_Activity_Hours,Health_Issues,Occupation,Smoking,Alcohol_Consumption
0,40,1,7,3.5,328.1,7.5,2.0,24.9,78,0.0,14.5,0.0,2,0,0
1,33,1,7,1.0,94.1,6.2,2.0,20.0,67,0.0,11.0,0.0,3,0,0
2,42,1,2,5.3,503.7,5.9,1.0,22.7,59,1.0,11.2,1.0,1,0,0
3,53,1,7,2.6,249.2,7.3,2.0,24.7,71,0.0,6.6,1.0,2,0,0
4,32,0,15,3.1,298.0,5.3,1.0,24.1,76,1.0,8.5,1.0,4,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9993,18,0,9,0.0,0.0,6.2,2.0,25.9,75,0.0,8.3,0.0,3,0,0
9994,49,0,7,1.6,150.1,5.7,1.0,25.5,81,1.0,12.9,1.0,4,0,1
9995,50,0,10,2.1,199.8,6.0,1.0,30.5,50,1.0,10.1,2.0,0,0,1
9996,18,0,18,3.4,319.2,5.8,1.0,19.1,71,1.0,11.6,1.0,3,0,0


- For **Gender, Country, and Occupation**: I used **Label Encoding** because these are nominal categorical features, meaning their values are just different categories without any natural order (e.g., Male, Female, Other or Egypt, USA, UK).  

- For **Sleep_Quality, Stress_Level, and Health_Issues**: I used **Ordinal Encoding** because these features have an inherent order.  
   - Sleep_Quality: Poor < Fair < Good < Excellent  
   - Stress_Level: Low < Medium < High  
   - Health_Issues: None < Mild < Moderate < Severe  
Ordinal encoding helps preserve this logical ranking when converting them into numbers.  


# Normalization

In [ ]:
scaler = StandardScaler()

df[numeric_features] = scaler.fit_transform(df[numeric_features])

- I applied **Normalization** to the numerical columns because their values have different ranges (for example: Age ranges from 18–80, Sleep_Hours from 3–10, and Caffeine_mg can go up to 1000).  

- Normalization rescales all numerical features into the same range, usually [0,1], so that no single feature dominates the analysis or the machine learning model.  

- This is especially useful when features are measured in different units (like hours, mg, and bpm), making the dataset more consistent and improving model performance.  
